radixdlt/src/main/java/com/radixdlt/store/berkeley/BerkeleySafetyStoreStore.java

```diff
@@ -0,0 +1,139 @@
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.radix.database.DatabaseEnvironment;
+
+import java.util.Objects;
+
+/**
+ * Store which persists state required to preserve the networks safety in case of a
+ * node restart.
+ *
+ * TODO: Prune saved safety state.
+ */
+public final class BerkeleySafetyStoreStore implements PersistentSafetyStateStore {
+	private static final String SAFETY_STORE_NAME = "safety_store";
+	private static final Logger logger = LogManager.getLogger();
+
+	private final DatabaseEnvironment dbEnv;
+	private Database safetyStore;
+
+	@Inject
+	public BerkeleySafetyStoreStore(DatabaseEnvironment dbEnv) {
+		this.dbEnv = Objects.requireNonNull(dbEnv, "dbEnv is required");
+
+		this.open();
+	}
+
+	private void fail(String message) {
+		logger.error(message);
+		throw new BerkeleyStoreException(message);
+	}
+
+	private void fail(String message, Exception cause) {
+		logger.error(message, cause);
+		throw new BerkeleyStoreException(message, cause);
+	}
+
+	private void open() {
+		DatabaseConfig primaryConfig = new DatabaseConfig();
+		primaryConfig.setAllowCreate(true);
+		primaryConfig.setTransactional(true);
+
+		try {
+			// This SuppressWarnings here is valid, as ownership of the underlying
+			// resource is not changed here, the resource is just accessed.
+			@SuppressWarnings("resource")
+			Environment env = this.dbEnv.getEnvironment();
+			this.safetyStore = env.openDatabase(null, SAFETY_STORE_NAME, primaryConfig);
+		} catch (Exception e) {
+			throw new BerkeleyStoreException("Error while opening database", e);
+		}
+
+		if (System.getProperty("db.check_integrity", "1").equals("1")) {
+			// TODO implement intergrity check
+		}
+	}
+
+	public Optional<Pair<Long, SafetyState>> get() {
+		try (com.sleepycat.je.Cursor cursor = this.safetyStore.openCursor(null, null)) {
+			DatabaseEntry pKey = new DatabaseEntry();
+			DatabaseEntry value = new DatabaseEntry();
+			OperationStatus status = cursor.getLast(pKey, value, LockMode.DEFAULT);
+			if (status == OperationStatus.SUCCESS) {
+				long lockedView = Longs.fromByteArray(value.getData());
+				long epochFound = ByteBuffer.wrap(pKey.getData()).getLong();
+				long view = ByteBuffer.wrap(pKey.getData()).getLong(Long.BYTES);
+
+				return Optional.of(Pair.of(epochFound, new SafetyState(View.of(view), View.of(lockedView))));
+			} else {
+				return Optional.empty();
+			}
+		}
+	}
+
+	@Override
+	public void commitState(Vote vote, SafetyState safetyState) {
+		if (!safetyState.getLastVotedView().equals(vote.getView())) {
+			throw new IllegalStateException("SafetyState and vote views don't match.");
+		}
+		long epoch = vote.getVoteData().getProposed().getLedgerHeader().getEpoch();
+		long view = vote.getView().number();
+		long lockedView = safetyState.getLockedView().number();
+		ByteBuffer keyByteBuffer = ByteBuffer.allocate(Long.BYTES * 2);

```

Comment:
Would be better to use arrays directly here — more code and more garbage produced by cycling through a ByteBuffer.

Something like:
```
    byte[] keyBytes = new byte[Long.BYTES * 3];
    Longs.copyTo(epoch, keyBytes, 0);
    Longs.copyTo(view, keyBytes, Long.BYTES);
```

Reply:
It also might have sense to change signature of `Long.toBytes(lon[] longArray)` to `Long.toBytes(long... longArray)`. This will enable to reduce code above to `byte[] keyBytes = Long.toBytes(epoch, view);`